add feature search engine adapater

mahdi00021 opened the pull request:
please check this "add feature search engine adapater"

ajmeese7 commented on src/search.js:
```diff
@@ -99,28 +102,18 @@ export default class Search {
         .catch(error => this.ui.emit('error', error));
     });
+    await Promise.all(this.adapters.map(a => a.init()));
+
```
FYI you have an extra new line here

ajmeese7 reviewed (commented):
Removing `package-lock.json` doesn't have an impact on the purpose of this PR, please restore the file so we can keep the changes focused on a single thing :)

mahdi00021 commented:
> Removing `package-lock.json` doesn't have an impact on the purpose of this PR, please restore the file so we can keep the changes focused on a single thing :)

@ajmeese7

I didn't touch the package-lock.json, I just checked and corrected the errors with eslint

andersevenrud commented:
> I didn't touch the package-lock.json

You did. You can see this in yhis PR if you go to the changes tab.

andersevenrud commented:
You also changed some test files. This should be reverted because it's out of scipe of this PR.

mahdi00021 commented:
> > I didn't touch the package-lock.json
> 
> You did. You can see this in this PR if you go to the changes tab.

Automatic eslint has changed

andersevenrud commented:
Eslint does not touch package lock. It also doesn't change the test files unless you tell it to do so.

So you must have done this by accident or something....

mahdi00021 commented:
> Eslint does not touch package lock. It also doesn't change the test files unless you tell it to do so.
> 
> So you must have done this by accident or something....

I commit now please check again

andersevenrud commented on package.json:
```diff
@@ -20,7 +20,7 @@
     "test": "npm run eslint && npm run stylelint && npm run jest",
     "jest": "jest",
     "coverage": "jest --coverage",
-    "eslint": "eslint index.js 'src/**/*.js'",
```
Revert this change

andersevenrud commented on src/adapters/search/vfs.js:
```diff
@@ -0,0 +1,31 @@
+            return [];
+          });
+      });
+    let k = Promise.all(promises).then(lists => [].concat(...lists));
```
Remove duplicate code and console logs.

mahdi00021 commented:
> 

Now I made the changes and corrections, there is nothing

andersevenrud commented:
Squash your commits and write a commit message that follows the conventional commits spesification, i.e.: `feat(search): adapter support`.

mahdi00021 commented:
> 

What should I do now?
delete branch again?  :)

andersevenrud commented:
You can use git interactive rebase to squash + change the commit message and do a force push.

No need to create new PR or branches. Git has the tools you need.

mahdi00021 commented:
> You can use git interactive rebase to squash + change the commit message and do a force push.
> 
> No need to create new PR or branches. Git has the tools you need.

please check this commit

andersevenrud commented:
You need to do a squash. This PR should only contain 1 commit.

mahdi00021 commented:
> You need to do a squash. This PR should only contain 1 commit.

please check

andersevenrud reviewed (commented):
You don't have to write "feature" in the message. The prefix `feat` already says this, so it's kind of redundant... but I can change that when merging.

I can't run this code until tomorrow because I'm not home. Maybe @ajmeese7 can have a look and approve. It's a shame there is no automated unit tests for search to verify everything works.

mahdi00021 commented:
> 

I tested it in my home system and it was executed successfully

andersevenrud commented:
> I tested it in my home system and it was executed successfully

But it needs to be *verified* by another human first since this stuff is not covered in automated tests.

mahdi00021 commented:
> > I tested it in my home system and it was executed successfully
> 
> But it needs to be _verified_ by another human first since this stuff is not covered in automated tests.

Can you do MERGE now?

andersevenrud commented:
No. Changes must be verified before it can be merged. I just explained why.

mahdi00021 commented:
> 



> No. Changes must be verified before it can be merged. I just explained why.

Do you have to do this?

andersevenrud commented:
Yes. No code gets merged before it has been verified either by another human or automated tests.

This is a standard process in programming.

mahdi00021 commented:
> Yes. No code gets merged before it has been verified either by another human or automated tests.
> 
> This is a standard process in programming.

So I will wait for you to accept merge

andersevenrud commented:
Can you do a git rebase onto the master branch?

I've added a unit test for the search service, and you have to run a rebase in order for this PR to run that test.

andersevenrud commented:
It would also be nice if you added a test case for adding custom adapters and ensure that they return results.

mahdi00021 commented:
> It would also be nice if you added a test case for adding custom adapters and ensure that they return results.

I added custom adapters for ensure that they return results in my system 
now what i do?

andersevenrud commented:
> now what i do?

First do a git rebase. On your branch:

```
git fetch origin master:master
git rebase master
git push --force
```

mahdi00021 commented:
> > now what i do?
> 
> First do a git rebase. On your branch:
> 
> ```
> git fetch origin master:master
> git rebase master
> git push --force
> ```

I do rebase , git said : Everything up-to-date

andersevenrud commented:
> Everything up-to-date

Oh... yes. You have a fork. So you need to add an `upstream` and use that instead of `origin`.

https://www.atlassian.com/git/tutorials/git-forks-and-upstreams

mahdi00021 commented:
> 

i add upstream your url or my fork url?

mahdi00021 commented:
> > Everything up-to-date
> 
> Oh... yes. You have a fork. So you need to add an `upstream` and use that instead of `origin`.
> 
> https://www.atlassian.com/git/tutorials/git-forks-and-upstreams

i do please check

andersevenrud commented:
FYI: Github sends out notifications whenever a PR is updated. You don't have to comment after you push changes.

---

You also should have gotten a notification yourself now that the tests have run, and failed.

So the next step now is to update the tests so that they pass.

Are you familiar with unit testing ?

mahdi00021 commented:
> FYI: Github sends out notifications whenever a PR is updated. You don't have to comment after you push changes.
> 
> You also should have gotten a notification yourself now that the tests have run, and failed.
> 
> So the next step now is to update the tests so that they pass.
> 
> Are you familiar with unit testing ?

Look, I'm confused, I changed everything you said, I put it on, it was successful, now you said do rebase, I did it, the test failed.

mahdi00021 commented:
> FYI: Github sends out notifications whenever a PR is updated. You don't have to comment after you push changes.
> 
> You also should have gotten a notification yourself now that the tests have run, and failed.
> 
> So the next step now is to update the tests so that they pass.
> 
> Are you familiar with unit testing ?

I'm confused, I changed everything you said, I put it on, it was successful, now you said do rebase, I did it, the test failed.
why?

Previously, the tests were successful

andersevenrud commented:
> Previously, the tests were successful

I added a **unit test** for the search so this could be automatically checked that it's working as intended.

You did a rebase, so **now** this test is running in this PR.

Since you changed the **API** of the search service, this test will fail. It needs to be updated to use the changes you introduced, which is the new `options` argument in the `Search` class.

If you look at the failed test that you got a notification for it says **exactly this**.

mahdi00021 commented:
> 

now Which test file should I change?

mahdi00021 commented:
> > Previously, the tests were successful
> 
> I added a **unit test** for the search so this could be automatically checked that it's working as intended.
> 
> You did a rebase, so **now** this test is running in this PR.
> 
> Since you changed the **API** of the search service, this test will fail. It needs to be updated to use the changes you introduced, which is the new `options` argument in the `Search` class.
> 
> If you look at the failed test that you got a notification for it says **exactly this**.

now Which test file should I change?
Search.js?

mahdi00021 commented:
> > Previously, the tests were successful
> 
> I added a **unit test** for the search so this could be automatically checked that it's working as intended.
> 
> You did a rebase, so **now** this test is running in this PR.
> 
> Since you changed the **API** of the search service, this test will fail. It needs to be updated to use the changes you introduced, which is the new `options` argument in the `Search` class.
> 
> If you look at the failed test that you got a notification for it says **exactly this**.



> > Previously, the tests were successful
> 
> I added a **unit test** for the search so this could be automatically checked that it's working as intended.
> 
> You did a rebase, so **now** this test is running in this PR.
> 
> Since you changed the **API** of the search service, this test will fail. It needs to be updated to use the changes you introduced, which is the new `options` argument in the `Search` class.
> 
> If you look at the failed test that you got a notification for it says **exactly this**.

Excuse me, how should I write a test? I am not familiar with the test writing of your project. In fact, my main expertise is PHP. I got to know Node a few weeks ago

andersevenrud commented:
> Excuse me, how should I write a test? I am not familiar with the test writing of your project. In fact, my main expertise is backend PHP python. I got to know Node a few weeks ago

You'd write a test just the same way as you would do for PHP or Python. The only difference here is the syntax, really.

It's only a matter of:

* Updating test with your new options argument
* Add a new test that checks that if you add another adapter, you get correct results back

https://github.com/os-js/osjs-client/blob/d6781762ce7948a5a9c19e16d1917f715b38de82/__tests__/search.js#L39

mahdi00021 commented:
> 



> > Excuse me, how should I write a test? I am not familiar with the test writing of your project. In fact, my main expertise is backend PHP python. I got to know Node a few weeks ago
> 
> You'd write a test just the same way as you would do for PHP or Python. The only difference here is the syntax, really.
> 
> It's only a matter of:
> 
>     * Updating test with your new options argument
> 
>     * Add a new test that checks that if you add another adapter, you get correct results back (optional...but I would love to see it)
> 
> 
> https://github.com/os-js/osjs-client/blob/d6781762ce7948a5a9c19e16d1917f715b38de82/__tests__/search.js#L39

how to pass custom adapter to test?

andersevenrud commented on __tests__/search.js:
```diff
@@ -39,15 +40,17 @@ class MockVFSServiceProvider extends ServiceProvider {
   let search;
-
+  let options = {};
+  options = MySearchAdapter;
```
This is wrong and not working. The API you added to the search class is `options = { adapters: [] }`

andersevenrud commented on src/adapters/search/custom-search.js:
```diff
@@ -0,0 +1,56 @@
+export default class MySearchAdapter {
```
This file does not belong here. If this is a test, just place it within the test file.

andersevenrud commented on src/adapters/search/custom-search.js:
```diff
@@ -0,0 +1,56 @@
+      }
+    ];
+
+    console.log('search:  ', obj);
```
No console statements in tests. Remove this.

andersevenrud commented on src/adapters/search/custom-search.js:
```diff
@@ -0,0 +1,56 @@
+    ];
+
+    console.log('search:  ', obj);
+    return Promise.all(obj)
```
This has no effect. You can do a straight `return obj`.

andersevenrud commented on src/adapters/search/custom-search.js:
```diff
@@ -0,0 +1,56 @@
+
+  async search(pattern) {
+
+    let obj = [
```
Use `const` for constants (not `let`).

andersevenrud reviewed (changes requested):
Looks like you've introduced some changes that does not belong here again:

* `CHANGELOG` should not be updated
* `package.json` should not be updated
* lint fixes under `__test__` not related to `search.js` should not be added

andersevenrud commented:
![add-feature-search-engine-adapater-by-mahdi00021-·-Pull-Request-202-·-os-js-osjs-client](https://user-images.githubusercontent.com/161548/210618901-d8adba19-4f2e-46a9-a04d-6815d5f223aa.png)

Not sure what you did, but now it's even more changes in here than before :sweat_smile: 

Are you working in an outdated repository ? If you create a an upstream (like discussed earlier) you can keep it up-to-date and then rebase your branch onto master branch to make sure it's in sync.

mahdi00021 commented:
> 



> ![add-feature-search-engine-adapater-by-mahdi00021-·-Pull-Request-202-·-os-js-osjs-client](https://user-images.githubusercontent.com/161548/210618901-d8adba19-4f2e-46a9-a04d-6815d5f223aa.png)
> 
> Not sure what you did, but now it's even more changes in here than before sweat_smile
> 
> Are you working in an outdated repository ? If you create a an upstream (like discussed earlier) you can keep it up-to-date and then rebase your branch onto master branch to make sure it's in sync.

I applay : 
git reset --soft HEAD~4 && git commit --no-verify -m "feat(search): add feature search adapter support"
git push --force origin featureSaerchEngine

for become to 1 commit and then be fails :((

mahdi00021 commented:
> ![add-feature-search-engine-adapater-by-mahdi00021-·-Pull-Request-202-·-os-js-osjs-client](https://user-images.githubusercontent.com/161548/210618901-d8adba19-4f2e-46a9-a04d-6815d5f223aa.png)
> 
> Not sure what you did, but now it's even more changes in here than before sweat_smile
> 
> Are you working in an outdated repository ? If you create a an upstream (like discussed earlier) you can keep it up-to-date and then rebase your branch onto master branch to make sure it's in sync.

I apply :
git reset --soft HEAD~4 && git commit --no-verify -m "feat(search): add feature search adapter support"
git push --force origin featureSaerchEngine

for become to 1 commit and then be fails :((

Added and changed several auto files

andersevenrud commented:
A `git reset` will *reset* your changes.

If you have multiple commits you want to join then use interactive rebase.

For *example*, if you have 10 commits: `git rebase -i HEAD~10`. Now, on the **bottom 9** change `pick` to `squash`.
This will squash the selected 9 onto the above making one commit in total. Then this can be force pushed.
